Add Cooperative Groups examples (#486)
amd/rocm-examples: HIP-Basic/cooperative_groups_double_buffered_tile/README.md

# Cooperative Groups Double-Buffered Tile Example

## Description

This program showcases a double-buffered tile load pipeline built from two cooperative groups

> APIs: the group-collective `cooperative_groups::memcpy_async` and the split barrier

Reviewer: is the term spilt barrier well understood? Should it be defined here at the start?

(`barrier_arrive` / `barrier_wait`) of a `thread_block`.

A single block streams a 1D array through two LDS (shared memory) buffers, one tile at a time.
The async load of the next tile is issued into the other buffer while the current tile is consumed,
and a split barrier separates the moment a thread has finished reading a buffer from the moment the

> block guarantees that every thread is done. The kernel applies the element-wise operation

Reviewer: ```suggestion
The async load of the next tile is issued into the second buffer while the current tile is consumed.
A split barrier separates the moment a thread has finished reading a buffer from the moment the
block guarantees that every thread is done. The kernel applies the element-wise operation
```

Reviewer: The second sentence could be a little more clear as to the split barrier separating the two moments: the moment the thread reads the buffer, and the moment all threads have completed the operation.

`out[i] = scale * in[i] + bias`, which is trivial to validate against a CPU reference.

`cooperative_groups::memcpy_async` is an **asynchronous**, group-collective copy (typically
global <-> LDS). HIP does not expose a separate wait handle (there is no `cg::wait()`), so its
completion must be enforced by a following group barrier - either a `block.sync()` (as the official
reference test does) or, as in this example, the `barrier_wait` of a split barrier whose
`barrier_arrive` is issued *after* the copy. Ordering matters: the prefetch of the next tile is
issued **before** `barrier_arrive`, so the release fence in `barrier_arrive` orders the copy's
completion and the acquire fence in `barrier_wait` makes the prefetched buffer visible to every
thread by the next iteration. The split barrier additionally lets the current tile's computation
run as independent work between `barrier_arrive` and `barrier_wait`. Correctness (no data races,
validated output) is the top priority.

This example targets the AMD/HIP (ROCm) backend: the NVIDIA path of
`<hip/cooperative_groups/memcpy_async.h>` is not yet implemented (the header carries a TODO for it),

> and it requires a ROCm version recent enough to ship that public header.

Reviewer: ```suggestion
This example targets the AMD/HIP (ROCm) backend,
and it requires a ROCm version recent enough to ship that public header.
```

Author: Thanks for catching, I was meaning to remove that and forgot.

### Application flow

> 1. A number of compile-time constants define the tile size (also the block size), the number of

Reviewer: ```suggestion
1. A number of compile-time constants define the tile size and block size, the number of
```

      calls `barrier_wait` to complete the barrier and the in-flight prefetch.
5. The result array is copied back to the host and all device memory is freed.
6. The elements of the result are compared with the CPU reference. The result of the comparison is
   printed to the standard output.

## Key APIs and Concepts

- `cooperative_groups::this_thread_block` returns the `thread_block` group that represents all
  threads of the block. The block is used both as the group for the collective copies and as the
  group that owns the split barrier.
- `cooperative_groups::memcpy_async(group, dst, src, count_in_bytes)` is an asynchronous
  group-collective copy that is designed for global <-> LDS transfers. Every thread of the group
  must call it collectively. **In HIP the `count` argument is expressed in bytes** (here `tile_size *
  sizeof(float)`). The copy is asynchronous and HIP exposes no separate wait handle, so its
  completion is enforced by a following group barrier (`block.sync()` or the split barrier's
  `barrier_wait`). On hardware or compilers without the asynchronous LDS builtins it falls back to a

>   correct traditional per-thread copy, so it always produces correct results.

Reviewer: ```suggestion
  sizeof(float)`). The copy is asynchronous and HIP exposes no separate wait handle, so the copy
  completion is enforced by a following group barrier (`block.sync()` or `barrier_wait`).
  On hardware or compilers without the asynchronous LDS builtins it falls back to a
  correct traditional per-thread copy, so it always produces correct results.
```

Reviewer: It is not clear to me what "it" refers to in "it falls back to"

> - The split barrier decomposes a block barrier into two phases. `thread_block::barrier_arrive`

Reviewer: This paragraph (or bullet) would be very helpful at the start of this topic for those who might be learning.

# Cooperative Groups Double-Buffered Tile Example

## Description

> This program showcases a double-buffered tile load pipeline built from two cooperative groups
> APIs: the group-collective `cooperative_groups::memcpy_async` and the split barrier of a
> `thread_block`. A split barrier decomposes an ordinary block barrier into two phases,
> `barrier_arrive` and `barrier_wait`, so that independent work can run between them instead of every
> thread blocking immediately.

Reviewer 2: ```suggestion
This program demonstrates double-buffered tile load pipeline built with two cooperative groups
APIs: 
* `cooperative_groups::memcpy_async` 
* The `thread_block` split barrier (`barrier_arrive()` / `barrier_wait()`)

A split barrier decomposes an ordinary block barrier into two phases:
* `barrier_arrive` 
* `barrier_wait`

This separation allows useful independent work to execute between the two operations instead of forcing every thread to block immediately at a single barrier.
```

A single block streams a 1D array through two LDS (shared memory) buffers, one tile at a time.
The async load of the next tile is issued into the second buffer while the current tile is consumed.
A split barrier separates the moment a thread has finished reading a buffer from the moment the
block guarantees that every thread is done. The kernel applies the element-wise operation

> `out[i] = scale * in[i] + bias`, which is trivial to validate against a CPU reference.

Reviewer: Is there a reference to the docs that can be added here?

Reviewer 3: Added a reference

> `cooperative_groups::memcpy_async` is an **asynchronous**, group-collective copy (typically
> global <-> LDS). HIP does not expose a separate wait handle (there is no `cg::wait()`), so its
> completion must be enforced by a following group barrier - either a `block.sync()` (as the official
> reference test does) or, as in this example, the `barrier_wait` of a split barrier whose
> `barrier_arrive` is issued *after* the copy. Ordering matters: the prefetch of the next tile is

Reviewer 2: ```suggestion
`cooperative_groups::memcpy_async()` is an **asynchronous**, group-collective copy operation, typically used to stage data between global memory and LDS (shared memory). HIP does not provide a separate completion handle or equivalent cg::wait() primitive, Therefore, completion of an asynchronous copy must be synchronized through a subsequent group barrier. The official reference test uses block.sync(), while this example relies on the `barrier_wait()` phase of a split barrier.

Correct ordering is critical. The prefetch of the next tile is
```

issued **before** `barrier_arrive`, so the release fence in `barrier_arrive` orders the copy's
completion and the acquire fence in `barrier_wait` makes the prefetched buffer visible to every
thread by the next iteration. The split barrier additionally lets the current tile's computation
run as independent work between `barrier_arrive` and `barrier_wait`. Correctness (no data races,
validated output) is the top priority.

> This example targets the AMD/HIP (ROCm) backend, and it requires a ROCm version recent enough to ship `hip/cooperative_groups/memcpy_async.h`.

Reviewer: You could mention a specific version, or later than...

Reviewer 3: Added
